### PR TITLE
Add template_from to node API links [OSF-4464]

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Allows a direct link between a node and it's template through and API link.

## Changes

Add properties to node model and link to API serializer.

## Side effects

None that I know of

## Ticket

https://openscience.atlassian.net/browse/OSF-4464

### CODE REVIEW

**Comment by brianjgeiger** [comment]
Tests would be awesome.

**Comment by brianjgeiger** [comment]
Pass complete. 🐧

**Comment by Johnetordoff** [comment]
@brianjgeiger Added tests, there's still small area of contention. If a registration is withdrawn the api will hide certain fields, should this be one of them? 

Currently registrations and draft_registrations are hidden, not sure what the rational behind this is so am just checking before I alter the test to that's failing to allow for the new field.

**Comment by brianjgeiger** [comment]
You might as well hide this field if it's withdrawn.

**Comment by Johnetordoff** [comment]
I was mistaken, all Relationship fields are hidden during withdrawal/reaction the test just didn't reflect this, now it does. Should be ready now.

**Comment by Johnetordoff** [comment]
@brianjgeiger this is failing a Phantom JS test, that shouldn't happen for something like this, right?

**Comment by brianjgeiger** [comment]
I've restarted the branch. If it fails again, you should look into it. Does it fail locally?

**Comment by Johnetordoff** [comment]
Yes, it also fails locally.

File: api/nodes/serializers.py [hunk 1]
```diff
@@ -164,6 +164,11 @@ class NodeSerializer(JSONAPISerializer):
         related_view_kwargs={'node_id': '<forked_from_id>'}
     )
 
+    templated_from = RelationshipField(
+        related_view=lambda n: 'registrations:registration-detail' if getattr(n, 'template_node', None) else 'nodes:node-detail',
```
Comment (brianjgeiger) [inline api/nodes/serializers.py]: I'm reasonably certain the getattr there should be grabbing 'is_registration' to determine if it's a registration or a node.

File: api_tests/base/test_serializers.py [hunk 2]
```diff
@@ -76,7 +76,7 @@ def test_registration_serializer(self):
         # fields that are visible for withdrawals
         visible_on_withdrawals = ['contributors', 'date_created', 'description', 'id', 'links', 'registration', 'title', 'type', 'current_user_can_comment']
         # fields that do not appear on registrations
-        non_registration_fields = ['registrations', 'draft_registrations']
+        non_registration_fields = ['registrations', 'draft_registrations', 'templated_from']
```
Comment (brianjgeiger) [inline api_tests/base/test_serializers.py]: Although if you're not showing the field at all on registrations, then you don't need the lambda on https://github.com/CenterForOpenScience/osf.io/pull/6038/files#r72799108 at all

File: api/nodes/serializers.py [hunk 3]
```diff
@@ -164,6 +164,11 @@ class NodeSerializer(JSONAPISerializer):
...
+        related_view_kwargs={'node_id': '<templated_from_id>'}
```
Comment (brianjgeiger) [inline api/nodes/serializers.py]: The gymnastics with the model are probably unnecessary. You can probably just do `'node_id': <'template_node._id>`

File: api_tests/base/test_serializers.py (hunk 2 above)
Comment (brianjgeiger) [inline api_tests/base/test_serializers.py]: I think if the model keeps the "template_node" on the registration, then it makes sense to display this on the registration, but it shouldn't be visible on withdrawals. So I think drop all the changes on this file and ensure it still works. You may have to change another test.

File: website/project/model.py [hunk 4]
```diff
@@ -2728,6 +2728,18 @@ def forked_from_id(self):
         return None
 
     @property
+    def templated_from(self):
```
Comment (brianjgeiger) [inline website/project/model.py]: As mentioned above, I don't _think_ you need to change the model at all to make this work. You should be able to drop the changes to this file.

**Comment by brianjgeiger** [comment]
Pass done. 🐧 I think this can be simplified, and I expect a registration test will fail after making some changes that I suggest and will need to be modified.

**Comment by brianjgeiger** [comment]
Looks good. 🎉 🐧 🎉

**Pull request merged**